Framework/Parallel/src/IO/EventParser.cpp

```diff
@@ -0,0 +1,199 @@
+#include "MantidParallel/IO/EventParser.h"
+#include "MantidTypes/TofEvent.h"
+#include <H5Cpp.h>
+#include <xmmintrin.h>
+
+using namespace Mantid::Parallel::IO;
+using Mantid::Types::TofEvent;
+
+namespace {
+template <class IndexType>
+std::pair<size_t, size_t>
+findStartAndEndPulses(const std::vector<IndexType> &eventIndex,
+                      size_t rangeStart, size_t count, size_t &curr) {
+  size_t startPulse = 0;
+  size_t endPulse = 0;
+
+  size_t pulse = curr;
+
+  const auto rangeEnd = rangeStart + count;
+
+  for (; pulse < eventIndex.size(); ++pulse) {
+    size_t icount =
+        (pulse != eventIndex.size() - 1 ? eventIndex[pulse + 1] : rangeEnd) -
+        eventIndex[pulse];
+    if (rangeStart >= static_cast<size_t>(eventIndex[pulse]) &&
+        rangeStart < static_cast<size_t>(eventIndex[pulse]) + icount)
+      startPulse = pulse;
+    if (rangeEnd > static_cast<size_t>(eventIndex[pulse]) &&
+        rangeEnd <= static_cast<size_t>(eventIndex[pulse]) + icount)
+      endPulse = pulse + 1;
+  }
+
+  curr = pulse;
+
+  return {startPulse, endPulse};
+}
+} // namespace
+
+namespace Mantid {
+namespace Parallel {
+namespace IO {
+
+template <class IndexType, class TimeZeroType, class TimeOffsetType>
+EventParser<IndexType, TimeZeroType, TimeOffsetType>::EventParser(
+    std::vector<std::vector<int>> rankGroups,
+    const std::vector<int32_t> &bankOffsets,
+    std::vector<std::vector<TofEvent> *> &eventLists)
+    : m_rankGroups(rankGroups), m_bankOffsets(bankOffsets),
+      m_eventLists(eventLists), m_posInEventIndex(0) {
+  if (m_bankOffsets.empty())
+    throw std::invalid_argument("The bankOffsets vector cannot be empty.");
+  if (m_eventLists.empty())
+    throw std::invalid_argument("The evenLists vector cannot be empty.");
```

Comment:
I am not sure such a check makes sense. You will be receiving input data with arbitrary `event_id` (or `bankIndex`), and unless you check every single one you have no sensible way of checking the length of `eventLists` (or `bankOffsets`). Having these to checks here gives a false sense of security, and having length 0 is no worse than having the length off by 1.

Maybe it makes more sense to put verbose warning comments at the appropriate points, saying that there is no range check, i.e., you just have to rely on the client using this class.